### DISCUSSION TITLE
Is there an example with a SQlite driver which works without Cgo

### DISCUSSION THREAD

**Discussion by thesobercoder** [ent/ent#1667] [Q&A]
I'm on a Windows 10 corporate laptop and I don't really want to install GCC. All the examples from the docs use the `github.com/mattn/go-sqlite3` which uses Cgo. Is there an example of a driver that doesn't use Cgo?

I tried using the `modernc.org/sqlite` package but am getting the following error - 

```go
package main

import (
	"context"
	"database/sql"
	"log"

	"entgo.io/ent/dialect"
	entsql "entgo.io/ent/dialect/sql"
	"github.com/thesobercoder/krongo/ent"
	"github.com/thesobercoder/krongo/ent/migrate"

	_ "modernc.org/sqlite"
)

func main() {
	db, err := sql.Open("sqlite", "file:db//test.db?mode=memory&cache=shared&_fk=1")
	if err != nil {
		log.Fatalf("failed opening connection to sqlite: %v", err)
	}
	defer db.Close()

	log.Println("Before open")
	drv := entsql.OpenDB(dialect.SQLite, db)

	log.Println("After open")

	client := ent.NewClient(ent.Driver(drv))
	if err != nil {
		log.Fatalf("failed connecting to sqlite: %v", err)
	}
	defer client.Close()

	log.Println("Before schema create")

	ctx := context.Background()
	err = client.Schema.Create(
		ctx,
		migrate.WithDropIndex(true),
		migrate.WithDropColumn(true),
		migrate.WithForeignKeys(true),
		migrate.WithFixture(true),
		migrate.WithGlobalUniqueID(true),
	)
	if err != nil {
		log.Fatalf("failed creating schema resources: %v", err)
	}
}
```

Error - 

```
failed creating schema resources: sql/schema: sqlite: foreign_keys pragma is off: missing "_fk=1" is the connection string
```

As you can see I've already passed in the `_fk=1` in the connection string, but it's still failing.

Any help please @a8m

**Comment by tux21b** [comment]
I just ran into the same problem. There are currently two issues:

a) ``github.com/mattn/go-sqlite3`` registers the name ``sqlite3`` which is also used by ``dialect.SQLite``. ``modernc.org/sqlite`` on the other hand registers the name ``sqlite`` by default. You can either create your database connection before (like you did) or register the modernc driver as ``sqlite3`` as well.

b) The second issue is the foreign key thing. Query parameters with a ``_`` prefix are not part of sqlite and are handled by  ``github.com/mattn/go-sqlite3`` only. The ``_fk`` parameter is one of them. I solved it by writing a custom driver that just calls the original driver and executes ``PRAGMA foreign_keys = on;`` on each new connection.

```
type sqliteDriver struct {
	*sqlite.Driver
}

func (d sqliteDriver) Open(name string) (driver.Conn, error) {
	conn, err := d.Driver.Open(name)
	if err != nil {
		return conn, err
	}
	c := conn.(interface{Exec(stmt string, args []driver.Value) (driver.Result, error)})
	if _, err := c.Exec("PRAGMA foreign_keys = on;", nil); err != nil {
		conn.Close()
		return nil, errors.Wrap(err, "failed to enable enable foreign keys")
	}
	return conn, nil
}

func init() {
        sql.Register("sqlite3", sqliteDriver{Driver: &sqlite.Driver{}})
}

func main() {
	client, err := ent.Open(dialect.SQLite, "file:ent?mode=memory&cache=shared")
	if err != nil {
		log.Fatalf("failed opening connection to sqlite: %v", err)
	}
	defer client.Close()

        // ...
}
```

**Reply by uffehellum** [reply]
Clean and elegant fix to other people's hard coded mistakes.

Thank you!!!

**Comment by xiaoqidun** [comment]
ent pure go sqlite3 driver, wrap modernc.org/sqlite
1. import _ "github.com/xiaoqidun/entps"
2. client, err := ent.Open("sqlite3", "file:./data.db")

**Comment by LittleKey** [comment]
just append `&_pragma=foreign_keys(1)` to ope uri

**Reply by xu756** [reply]
> &_pragma=foreign_keys(1)

good

**Reply by spekary** [reply]
This will not necessarily work. The pragma only applies to the current connections in the connection pool. Since new connections can be blindly created by the driver, those new connections will NOT have the pragma.

**Reply by Nintron27** [reply]
> The pragma only applies to the current connections in the connection pool. ...new connections will NOT have the pragma.

Am I perhaps missing something? I tested adding new connections to the pool and every single new one has the PRAGMAs enabled as set in the DSN.

**Comment by spekary** [comment]
Try this:

```
func init() {
	// this is required to enable foreign key checks on every connection, even ones that are created blindly
	// by the driver's connection pool mechanism.
	sqlite.RegisterConnectionHook(func(conn sqlite.ExecQuerierContext, dsn string) error {
		// enable FK enforcement (or any other PRAGMAs you like)
		_, err := conn.ExecContext(context.Background(), "PRAGMA foreign_keys = ON", nil)
		return err
	})
}

```

**Comment by godcong** [comment]
```
import (
    _ "https://github.com/sqlite3ent/sqlite3"
)
```

It's use `modernc.org/sqlite` and merged `github.com/mattn/go-sqlite3` init actions.
